File: AlgoExpert/Arrays/07-Three-Number-Sum/Java/Solution.java
import java.util.*;

class Solution {
  public static List<Integer[]> threeNumberSum(int[] array, int targetSum) {
    // Write your code here.
    if (array.length < 3) {
      return new ArrayList<Integer[]>();
    }
    List<Integer[]> triplets = new ArrayList<>();
    
    Arrays.sort(array); // O(nlogn)

    for (int i = 0; i < array.length; i++) {
      int currValue = array[i];
      int leftPointer = i + 1;
      int rightPointer = array.length - 1;
      
      while (leftPointer < rightPointer) {
        int currSum = currValue + array[leftPointer] + array[rightPointer];
        
        if (currSum == targetSum) {
          triplets.add(new Integer[]{currValue, array[leftPointer], array[rightPointer]});
          leftPointer += 1;
          rightPointer -= 1;
        }
        else if (currSum < targetSum) {
          leftPointer += 1;
        }
        else if (currSum > targetSum) {
          rightPointer -= 1;
        }
      }
    }
		
    return triplets;
  }
}

File: AlgoExpert/Arrays/07-Three-Number-Sum/Python/solution.py
def threeNumberSum(array, targetSum):
    # Write your code here.
    if len(array) < 3:
        return []

    triplets = list()
    array.sort()

    for i in range(len(array)):
        curr_value = array[i]
        left_pointer = i + 1
        right_pointer = len(array) - 1

        while left_pointer < right_pointer:
            curr_sum = curr_value + array[left_pointer] + array[right_pointer]

            if curr_sum == targetSum:
                triplet = [curr_value, array[left_pointer], array[right_pointer]]
                triplets.append(triplet)
                left_pointer += 1
                right_pointer  -= 1
            elif curr_sum < targetSum:
                left_pointer += 1
            elif curr_sum > targetSum:
                right_pointer  -= 1
                
    return triplets
